stm32/libcanard/bxcan/src/bxcan.c

```diff
@@ -0,0 +1,753 @@
+                        tx_mailbox = 0xFF;
+                        break;
+                    }
+                }
+            }
+
+            if (tx_mailbox == 0xFF)
+            {
+                // All TX mailboxes are busy (this is highly unlikely); at the same time we know that there is no
+                // higher or equal priority frame that is currently pending. Therefore, priority inversion has
+                // just happend (sic!), because we can't enqueue the higher priority frame due to all TX mailboxes
+                // being busy. This scenario is extremely unlikely, because in order for it to happen, the application
+                // would need to transmit 4 (four) or more CAN frames with different CAN ID ordered from high ID to
+                // low ID nearly at the same time. For example:
+                //  1. 0x123        <-- Takes mailbox 0 (or any other)
+                //  2. 0x122        <-- Takes mailbox 2 (or any other)
+                //  3. 0x121        <-- Takes mailbox 1 (or any other)
+                //  4. 0x120        <-- INNER PRIORITY INVERSION HERE (only if all three TX mailboxes are still busy)
+                // This situation is even less likely to cause any noticeable disruptions on the CAN bus. Despite that,
+                // it is better to warn the developer about that during debugging, so we fire an assertion failure here.
+                // It is perfectly safe to remove it.
+                // Note: this error condition (all TX mailboxes busy) also occurs when erroneously sending data to the
+                // TX mailboxes of a non-existing CAN interface. Eg: accessing CAN2 on a device that only has CAN1.
+#if BXCAN_CONFIG_DEBUG_INNER_PRIORITY_INVERSION
+                BXCAN_ASSERT(!"CAN PRIO INV");  // NOLINT
+#endif
+                prio_inversion = true;
+            }
+            else
+            {
+                prio_inversion = false;
+            }
+        }
+        else  // All TX mailboxes are free, use first
+        {
+            tx_mailbox = 0U;
+        }
+
+        BXCAN_ASSERT(tx_mailbox < 3U);  // Index check - the value must be correct here
+    }
+
+    // By this time we've proved that a priority inversion would not occur, and we've also
+    // found a free TX mailbox. Therefore it is safe to enqueue the frame now.
+    if ((input_ok == true) && (prio_inversion == false))
+    {
+        volatile BxCANTxMailboxType* const mb = &bxcan_base->TxMailbox[tx_mailbox];
+
+        // The payload can be any length up to the MTU. We first make a local copy of the
+        // available payload bytes in a zero-filled array of MTU size. This makes the logic
+        // for filling the mailbox simpler as the payload pointer can be to variable-size
+        // data or even be a NULL pointer.
+        static uint8_t scratch_data[8] = {0};  // Not strictly needed to zero the storage.
```

Comment:
One would not expect a scratch buffer to introduce a race condition.

```suggestion
        uint8_t scratch_data[8] = {0};  // Not strictly needed to zero the storage.
```

Reply:
Given the other places where I'm taking undue liberties with static, can you point me to a good discussion/tutorial on this in the context of embedded programming?

Reply:
IIRC Kernighan & Ritchie give this question an adequate treatment. The gist is that a statically allocated object exists in one copy from the moment the program begins its execution (before main() is entered) (in C++ this bit is somewhat more involved if the constructor of the static object is non-trivial) until it is terminated (either by returning from main() or abnormally). It follows then that if there are multiple execution contexts (threads, interrupts, etc.) accessing the same static mutable (non-const) object, a race condition may occur. To avoid that, static storage should only be used when a shared state is required by the business logic.

The above is only a small slice of the broader subject of mutability and shared state.

Note that it is not related to `static const` because in this case, the affected data is inherently immutable (hence const) and therefore is not prone to causing a race condition:

```c
static const int PI = 3;  // Just a const, fine.
static int       pi = 3;  // Danger
int              pi = 3;  // A context-local (stack-allocated) variable, fine.
```

Reply:
I ordered the book. All I have on C are too basic to cover this.

Ok, so if I get this, the moment I declare, say, a "static int" someplace in a routine, this effectively becomes a persistent, global variable that is accessible and mutable from any place. That is scary business indeed.

Reply:
> accessible and mutable from any place

Not from any place but from the context of its declaration, but it doesn't help much.

This driver is not reentrant, meaning that it is not supposed to be accessed from multiple contexts concurrently, but it doesn't mean that statics introduce no cost.

Reply:
So, a situation where this would go horribly wrong is if two threads were to access the driver. In that case, the static variables would be shared from both threads. With a context-local, the streams would not cross in that way. If this is closer to the mark.

What is the cost of static vs stack-allocated?

Reply:
By cost I mean risk. Performance-wise it could make a difference in cached processors because the stack memory is usually hot and available in L1, whereas static variables may reside in different segments that may be slow to access.

There is an excellent textbook that covers this and related questions. I read it in the [original language](https://www.rahvaraamat.ee/p/%D1%81%D0%B5%D1%82%D0%B5%D0%B2%D1%8B%D0%B5-%D0%BE%D0%BF%D0%B5%D1%80%D0%B0%D1%86%D0%B8%D0%BE%D0%BD%D0%BD%D1%8B%D0%B5-%D1%81%D0%B8%D1%81%D1%82%D0%B5%D0%BC%D1%8B/58072/en?isbn=9785911805289#) but there appears to be [an edition in English](https://www.amazon.com/Computer-Networks-Principles-Technologies-Protocols/dp/0470869828) as well.

Reply:
I checked the links. The English book is the author's other book on computer networking, but not the one on networked operating systems. As far as I could find, there appears to be no English version of the relevant one. (Also, surprising how well computer translation works these days. It had no trouble with the Russian book description.)